CI: move CI to github

mkalcok opened the pull request:
This PR facilitates move from opendev to github. It fixes various issues that were broken with the building/linting/unit testing/functional testing. It also contains new github workflows that will take over functionality of the old CI system.

Please review individual commits.

mkalcok committed:
layer: Switch repo reference to github.

We are moving the charm development to github.

Signed-off-by: Martin Kalcok <martin.kalcok@canonical.com>

mkalcok committed:
charmcraft: Bump rust toolchain to 1.80.

Python package 'jsonschema' depends on rust bindings (rpds package)
to compile. rpds, in turn, uses rust's lockfile v4, which became
supported in cargo 1.78 [0].

[0] https://releases.rs/docs/1.78.0/

Signed-off-by: Martin Kalcok <martin.kalcok@canonical.com>

mkalcok committed:
tox: Define zaza constraint file for func. tests.

Constraint file allows control over which juju library is
used by zaza in functional tests. This change ensures that
functional tests run, using Juju 3.

Signed-off-by: Martin Kalcok <martin.kalcok@canonical.com>

mkalcok committed:
test: switch principal charm from magpie to ubuntu

Since charm-ovn-chassis is a subordinate charm, it needs a principal
charm to deploy to. This used to be 'magpie' but that charm
doesn't have a 'noble' series release. To allow us to test ovn-chassis
noble release, we are switching the principal charm to 'ubuntu'.

func-test-pr: https://github.com/openstack-charmers/zaza-openstack-tests/pull/1308
Signed-off-by: Martin Kalcok <martin.kalcok@canonical.com>

mkalcok committed:
pep8: Bump flake8 package to 7.1.1

This fixes compatibility issues between flake8 plugins and
previously used flake8 3.9.2.

Signed-off-by: Martin Kalcok <martin.kalcok@canonical.com>

fnordahl commented on src/tests/bundles/noble-caracal.yaml:
```diff
@@ -20,19 +20,11 @@ applications:
-    # instance vCPUs requested.
-    #
-    # Let's use a 8 VCPU 2 socket flavor for low level tests
-    constraints: "instance-type=twosocketm1.xlarge"
```
Nice to see that we don't need this anymore, this resonates with my finding that LXD constructs CPU topologies inside the guest that resemble what you would expect from a physical system, as opposed to the OpenStack default of constructing a multi socket CPU toplogy, with number of CPU sockets being the same as the number of VCPUs.

This detail matters because Open vSwitch DPDK memory allocation is influenced by the number of CPU sockets.

This should be sufficiently documented in the test code, so anyone attempting this on a multi socket system should get enough clues for where to look.

mkalcok replied:
> This should be sufficiently documented in the test code, so anyone attempting this on a multi socket system should get enough clues for where to look.

Should I bring back the comment that was originally here, or do you mean that it should be documented in the `zaza-openstack-tests` code?

fnordahl replied:
Ah, no action needed, the intention with the comment was to say it is already sufficiently documented in the test code, ref:
https://github.com/openstack-charmers/zaza-openstack-tests/blob/dd14fe7abb1b7caacb480b4ef407a8acac555822/zaza/openstack/charm_tests/test_utils.py#L646-L684

fnordahl commented on src/tests/bundles/noble-caracal.yaml:
```diff
@@ -20,19 +20,11 @@ applications:
-    options:
-      source: *openstack-origin
     channel: latest/edge
+    constraints: "virt-type=virtual-machine mem=8G"
```
This would likely mean Juju requesting a 1 VCPU instance. Is this really sufficient?

mkalcok replied:
for some reason I thought that it default's to 2 vCPUs. Functional tests were passing OK, so I'd say that it is "sufficient", but I'll try locally if increasing vCPU count gives us some meaningful performance boost. We should have 4 CPU's to work with in the GH runner.

mkalcok replied:
Just for the record, adding `cores=2` constraint shaved of only 1 seconds of 10min run on my machine, so It's probably not worth it to commit all available cores to the test VMs.

mkalcok commented:
I'm marking this PR as draft until we figure out how to trigger github workflwos. My hunch is that we'll have to marge the workflows first into the default branch #3

mkalcok committed:
CI: Reference GH actions by full hash.

Github CI refuses to run Github Actions that are referenced only
via the short hash.

Signed-off-by: Martin Kalcok <martin.kalcok@canonical.com>

mkalcok committed:
CI: Install juju-crashdump when needed.

Signed-off-by: Martin Kalcok <martin.kalcok@canonical.com>

fnordahl reviewed (approved):
LGTM, thank you for your work on this!
